apps/test/unit/promiseChainTest.js

```diff
@@ -0,0 +1,225 @@
+    sequence.push('C');
+    expect(sequence).to.deep.equal(['A', 'B', 'C']);
+  });
+
+  it(`and when creating new Promises this doesn't seem like an issue`, () => {
+    // because a new promise's callback gets called immediately
+    const _ = new Promise(resolve => {
+      sequence.push('A'); // Invoked immediately
+      setTimeout(() => {
+        sequence.push('C');
+        resolve();
+      }, 0);
+    });
+    sequence.push('B');
+    clock.tick(0); // Calls the scheduled timeout _immediately_
+    sequence.push('D');
+    expect(sequence).to.deep.equal(['A', 'B', 'C', 'D']);
+  });
+
+  it(`but Promise.then doesn't happen immediately, even after Promise.resolve()`, done => {
+    Promise.resolve()
+        .then(() => {
+          sequence.push('C');
+          expect(sequence).to.deep.equal(['A', 'B', 'C']);
+          done();
+        });
+    sequence.push('A');
+    clock.tick(0); // Does _not_ cause Promise.then to occur
+    sequence.push('B');
+  });
+
+  it(`which can lead to some unexpected results.`, done => {
+    Promise.resolve().then(() => {
+      sequence.push('C');
+      setTimeout(() => {
+        sequence.push('E');
+        done();
+      }, 0);
+      sequence.push('D');
+      clock.tick(0); // So we have to tick again inside the promise...
+      expect(sequence).to.deep.equal(['A', 'B', 'C', 'D', 'E']);
+    });
+    sequence.push('A');
+    clock.tick(0); // Does nothing because tick has not been scheduled!
+    sequence.push('B');
+  });
+
+  it(`Promise "then" calls seem to be queued in the order they are received`, done => {
+    Promise.resolve().then(() => {
+      sequence.push('B');
+    });
+    sequence.push('A');
+    Promise.resolve().then(() => {
+      sequence.push('C');
+      expect(sequence).to.deep.equal(['A', 'B', 'C']);
+      done();
+    });
+  });
+
+  it(`But in complex situations this gets very hard to reason about`, done => {
+    const promiseChainBeingTested = Promise.resolve()
+        .then(() => {
+          return new Promise(resolve => setTimeout(() => {
+            sequence.push('A');
+            resolve();
+          }, 1));
+        })
+        .then(() => {
+          return new Promise(resolve => setTimeout(() => {
+            sequence.push('B');
+            resolve();
+          }, 1));
+        })
+        .then(() => {
+          sequence.push('C');
+        });
+
+    sequence.push('_');
+
+    const testCode = Promise.resolve()
+        .then(() => {
+          sequence.push('a');
+          clock.tick(1);
+        })
+        .then(() => {
+          sequence.push('b');
+        })
+        .then(() => {
+          sequence.push('c');
+          clock.tick(1);
+        })
+        .then(() => {
+          sequence.push('d');
+        });
+
+    Promise.all([promiseChainBeingTested, testCode]).then(() => {
+      expect(sequence).to.deep.equal(['_', 'a', 'A', 'b', 'c', 'B', 'd', 'C']);
+      done();
+    }).catch(done);
+  });
```

Comment:
I don't see the value in having this test show that promise behavior is "confusing". I'd rather have this test disprove the premise that promises are executed in FIFO order, while also showing exactly what the expected and actual sequence of events is. Here is a shorter example:
```js
  it(`But promises don't always get queued in the order you expect`, done => {
    const promiseChainBeingTested = Promise.resolve()
        .then(() => {
          sequence.push(`A`);
          return new Promise(resolve => setTimeout(() => {
            sequence.push('B');
            resolve();
          }, 1));
        })
        .then(() => {
          sequence.push('C');
        });

    sequence.push('_');

    sequence.push('a');
    const testCode = Promise.resolve()
        .then(() => {
          sequence.push(`b`);
          clock.tick(1);
          sequence.push('c');
        })
        .then(() => {
          sequence.push('d');
        })
        .then(() => {
          sequence.push('e');
        });


    Promise.all([promiseChainBeingTested, testCode]).then(() => {
      // The order one might expect if promises were FIFO is:
      // _, a, A, b, B, c, C, d, e
      expect(sequence).to.deep.equal(['_', 'a', 'A', 'b', 'B', 'c', 'd', 'C', 'e']);
      done();
    }).catch(done);
  });
```